File: web-app/src/logic/sorters.js
const getSortableArtist = (name) => {
  if (!name) return "";
  const n = String(name).trim();
  if (n.toLowerCase().startsWith("the ")) {
    return n.slice(4).trim();
  }
  return n;
};

const getPriorityAddedDate = (album) => {
  const t = album.tags || {};
  const yt = parseInt(t.UNIX_ADDED_YOUTUBE) || 0;
  if (yt > 0) return yt;

  const am = parseInt(t.UNIX_ADDED_APPLEMUSIC) || 0;
  if (am > 0) return am;

  const fb = parseInt(t.UNIX_ADDED_FOOBAR) || 0;
  if (fb > 0) return fb;

  return parseInt(album.unix_added) || 0;
};

export const sorters = {
  default: (a, b) => {
    const artistA = getSortableArtist(a.tags?.CUSTOM_ALBUMARTIST || a.ALBUMARTIST).toLowerCase();
    const artistB = getSortableArtist(b.tags?.CUSTOM_ALBUMARTIST || b.ALBUMARTIST).toLowerCase();
    const artistComp = artistA.localeCompare(artistB);
    if (artistComp !== 0) return artistComp;

    const dateA = a.DATE || "0000";
    const dateB = b.DATE || "0000";
    const dateComp = dateA.localeCompare(dateB);
    if (dateComp !== 0) return dateComp;

    const titleA = (a.ALBUM || "").toLowerCase();
    const titleB = (b.ALBUM || "").toLowerCase();
    return titleA.localeCompare(titleB);
  },

  date_added: (a, b) => getPriorityAddedDate(b) - getPriorityAddedDate(a),
  
  az: (a, b) => (a.ALBUM || "").localeCompare(b.ALBUM || ""),
  
  artist: (a, b) => {
    const artA = a.tags?.CUSTOM_ALBUMARTIST || a.ALBUMARTIST || "";
    const artB = b.tags?.CUSTOM_ALBUMARTIST || b.ALBUMARTIST || "";
    return artA.localeCompare(artB);
  },
  
  year: (a, b) => {
    const dateA = a.DATE || "0000";
    const dateB = b.DATE || "0000";
    return dateB.localeCompare(dateA);
  },

  entropy: (a, b) => (b.tags?.COVER_ENTROPY || 0) - (a.tags?.COVER_ENTROPY || 0),

  chroma: (a, b) => (b.tags?.COVER_CHROMA || 0) - (a.tags?.COVER_CHROMA || 0),

  duration: (a, b) => (b.album_duration || 0) - (a.album_duration || 0)
};

export const SORTER_LABELS = {
  default: "Default",
  az: "Alphabetical",
  year: "Year",
  date_added: "Date Added",
  duration: "Duration",
  chroma: "Chroma",
  entropy: "Entropy",
};

File: rust/src/egui/logic/sorters.rs
use crate::server::library::models::AlbumView;
use std::cmp::Ordering;

fn get_sortable_artist(name: &str) -> String {
    let n = name.trim();
    if n.to_lowercase().starts_with("the ") {
        n[4..].trim().to_lowercase()
    } else {
        n.to_lowercase()
    }
}

pub fn sort_default(a: &AlbumView, b: &AlbumView) -> Ordering {
    let artist_a = get_sortable_artist(&a.album_data.albumartist);
    let artist_b = get_sortable_artist(&b.album_data.albumartist);
    let artist_cmp = artist_a.cmp(&artist_b);
    if artist_cmp != Ordering::Equal { return artist_cmp; }

    let date_a = &a.album_data.date;
    let date_b = &b.album_data.date;
    let date_cmp = date_a.cmp(date_b);
    if date_cmp != Ordering::Equal { return date_cmp; }

    a.album_data.album.to_lowercase().cmp(&b.album_data.album.to_lowercase())
}

pub fn sort_date_added(a: &AlbumView, b: &AlbumView) -> Ordering {
    b.album_data.info.unix_added.cmp(&a.album_data.info.unix_added)
}

#[allow(dead_code)]
pub fn sort_alphabetical(a: &AlbumView, b: &AlbumView) -> Ordering {
    a.album_data.album.to_lowercase().cmp(&b.album_data.album.to_lowercase())
}
